docs/writing-your-first-rclrs-node.md

```diff
@@ -0,0 +1,226 @@
+  <name>republisher_node</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="user@todo.todo">user</maintainer>
+  <license>TODO: License declaration</license>
+
+  <depend>rclrs</depend>
+  <depend>std_msgs</depend>
+
+  <export>
+    <build_type>ament_cargo</build_type>
+  </export>
+</package>
+```
+
+
+## Writing the basic node structure
+
+Since Rust doesn't have inheritance, it's not possible to inherit from `Node` as is common practice in `rclcpp` or `rclpy`.
+
+Instead, you can store the node as a regular member. Let's add a struct that contains the node, a subscription, and a field for the last message that was received to `main.rs`:
+
+```rust
+use std::sync::Arc;
+use std_msgs::msg::String as StringMsg;
+
+struct RepublisherNode {
+    node: rclrs::Node,
+    _subscription: Arc<rclrs::Subscription<StringMsg>>,
+    data: Option<StringMsg>,
+}
+
+impl RepublisherNode {
+    fn new(context: &rclrs::Context) -> Result<Self, rclrs::RclrsError> {
+        let mut node = rclrs::Node::new("republisher", context)?;
+        let data = None;
+        let _subscription = node.create_subscription(
+            "in_topic",
+            rclrs::QOS_PROFILE_DEFAULT,
+            |msg: StringMsg| { todo!("Assign msg to self.data") },
+        )?;
+        Ok(Self {
+            node,
+            _subscription,
+            data,
+        })
+    }
+}
+```
+
+Next, add a main function to launch it:
+
+```rust
+fn main() -> Result<(), rclrs::RclrsError> {
+    let context = rclrs::Context::new(std::env::args())?;
+    let republisher = RepublisherNode::new(&context)?;
+    rclrs::spin(&republisher.node)
+}
+```
+
+You should now be able to run this node with `cargo run`. However, the subscription callback still has a `todo!` in it, so it will exit with an error when it receives a message.
+
+
+## Storing received data in the struct
+
+Let's do something about that `todo!`. The obvious thing for the subscription callback to do would be this:
+
+```rust
+|msg: StringMsg| {
+    data = Some(msg);
+},
+```
+
+This is a standard pattern in C++, but doesn't work in Rust. Why not?
+
+Written like this, `data` is *borrowed* by the callback, but `data` is a local variable only exists in its current form until the end of `RepublisherNode::new()`. The subscription callback is required to not borrow any variables, because the subscription, and therefore the callback, can live indefinitely.
+
+> 💡 As an aside, this requirement is expressed by the `'static` bound on the generic parameter `F` for the callback in `Node::create_subscription()`.
+
+You might think "I don't want to borrow from the local variable `data` anyway, I want to borrow from the `data` field in `RepublisherNode`!" and you would be right. That variable lives considerably longer, but also not forever, so it wouldn't be enough. A secondary problem is that it would be a *self-referential struct*, which is not allowed in Rust.
```

Comment:
Would it be useful for people seeing this to be directed to [this series](https://rust-unofficial.github.io/too-many-lists/) as to why self-referential structs (and trying to squeeze them into Rust in any way the type system will let you) is not a good thing?